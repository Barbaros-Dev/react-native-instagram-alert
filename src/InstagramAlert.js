import React, { Component } from 'react';
import {
  Text,
  Animated,
  View,
  TouchableOpacity,
  TouchableWithoutFeedback,
  ActivityIndicator,
  BackHandler,
  Modal,
  Platform,
} from 'react-native';

import PropTypes from 'prop-types';

import styles from './styles';

const HwBackHandler = BackHandler;
const HW_BACK_EVENT = 'hardwareBackPress';

const { OS } = Platform;

export default class InstagramAlert extends Component {
  constructor(props) {
    super(props);
    const { show } = this.props;
    this.springValue = new Animated.Value(props.animatedValue);

    this.state = {
      showSelf: false,
    };

    if (show) this._springShow(true);
  }

  componentDidMount() {
    HwBackHandler.addEventListener(HW_BACK_EVENT, this._handleHwBackEvent);
  }

  _springShow = (fromConstructor) => {
    const { useNativeDriver = false } = this.props;

    this._toggleAlert(fromConstructor);
    Animated.spring(this.springValue, {
      toValue: 1,
      bounciness: 10,
      useNativeDriver,
    }).start();
  };

  _springHide = () => {
    const { useNativeDriver = false } = this.props;

    if (this.state.showSelf === true) {
      Animated.spring(this.springValue, {
        toValue: 0,
        tension: 10,
        useNativeDriver,
      }).start();

      setTimeout(() => {
        this._toggleAlert();
        this._onDismiss();
      }, 70);
    }
  };

  _toggleAlert = (fromConstructor) => {
    if (fromConstructor) this.state = { showSelf: true };
    else this.setState({ showSelf: !this.state.showSelf });
  };

  _onDismiss = () => {
    const { onDismiss } = this.props;
    onDismiss && onDismiss();
  };

  _renderAlert = () => {
    const animation = { transform: [{ scale: this.springValue }] };

    const { showProgress } = this.props;
    const { title, message, customView = null } = this.props;

    const {
      showCancelButton,
      cancelText,
      cancelButtonStyle,
      cancelButtonTextStyle,
      onCancelPressed,
      cancelButtonTestID
    } = this.props;

    const {
      showConfirmButton,
      confirmText,
      confirmButtonStyle,
      confirmButtonTextStyle,
      onConfirmPressed,
      confirmButtonTestID
    } = this.props;

    const {
      alertContainerStyle,
      overlayStyle,
      progressSize,
      progressColor,
      contentContainerStyle,
      contentStyle,
      titleStyle,
      messageStyle,
      actionContainerStyle,
    } = this.props;

    const cancelButtonData = {
      testID: cancelButtonTestID,
      text: cancelText,
      buttonStyle: cancelButtonStyle,
      buttonTextStyle: cancelButtonTextStyle,
      onPress: onCancelPressed,
    };

    const confirmButtonData = {
      testID: confirmButtonTestID,
      text: confirmText,
      buttonStyle: confirmButtonStyle,
      buttonTextStyle: confirmButtonTextStyle,
      onPress: onConfirmPressed,
    };

    return (
      <View style={[styles.container, alertContainerStyle]}>
        <TouchableWithoutFeedback onPress={this._onTapOutside}>
          <View style={[styles.overlay, overlayStyle]} />
        </TouchableWithoutFeedback>
        <Animated.View
          style={[styles.contentContainer, animation, contentContainerStyle]}
        >
          <View style={[styles.content, contentStyle]}>
            {title ? (
              <Text style={[styles.title, titleStyle]}>{title}</Text>
            ) : null}
            {message ? (
              <Text style={[styles.message, messageStyle]}>{message}</Text>
            ) : null}
            {customView}
          </View>
          <View style={{ backgroundColor: "#EFEFEF", position: "absolute", width: 258, height: 1, bottom: 95 }}></View>
          <View style={{ backgroundColor: "#EFEFEF", position: "absolute", width: 258, height: 1, bottom: 48 }}></View>
          <View style={[styles.action, actionContainerStyle]}>
            
            <TouchableOpacity onPress={confirmButtonData.onPress} style={{ backgroundColor: "#fff", width: 258, height: 47, position: "absolute", bottom: 48, justifyContent: "center", alignItems: "center" }}>
              <Text style={{ color: "#FF4D57" }}>{confirmButtonData.text}</Text>
            </TouchableOpacity>
            <View style={{ backgroundColor: "#EFEFEF", position: "absolute", width: 258, height: 1, bottom: 48 }}></View>
            <TouchableOpacity onPress={cancelButtonData.onPress} style={{ backgroundColor: "#fff", width: 258, height: 47, position: "absolute", bottom: 0, borderBottomLeftRadius: 20, borderBottomRightRadius: 20, justifyContent: "center", alignItems: "center" }}>
              <Text>{cancelButtonData.text}</Text>
            </TouchableOpacity>
          </View>
        </Animated.View>
      </View>
    );
  };

  render() {
    const { show, showSelf } = this.state;
    const { modalProps = {}, closeOnHardwareBackPress } = this.props;

    const wrapInModal = OS === 'android' || OS === 'ios';

    return showSelf ?
      wrapInModal ? (
        <Modal
          animationType="none"
          transparent={true}
          visible={show}
          onRequestClose={() => {
            if (showSelf && closeOnHardwareBackPress) {
              this._springHide();
            }
          }}
          {...modalProps}
        >
          {this._renderAlert()}
        </Modal>
      ) : this._renderAlert()
    : null;
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    const { show } = nextProps;
    const { showSelf } = this.state;

    if (show && !showSelf) this._springShow();
    else if (show === false && showSelf) this._springHide();
  }

  componentWillUnmount() {
    HwBackHandler.removeEventListener(HW_BACK_EVENT, this._handleHwBackEvent);
  }
}

InstagramAlert.propTypes = {
  show: PropTypes.bool,
  animatedValue: PropTypes.number,
  useNativeDriver: PropTypes.bool,
  title: PropTypes.string,
  message: PropTypes.string,
  showCancelButton: PropTypes.bool,
  showConfirmButton: PropTypes.bool,
  cancelText: PropTypes.string,
  confirmText: PropTypes.string,
  onCancelPressed: PropTypes.func,
  onConfirmPressed: PropTypes.func,
  customView: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.node,
    PropTypes.func,
  ]),
  modalProps: PropTypes.object,
  cancelButtonTestID: PropTypes.string,
  confirmButtonTestID: PropTypes.string
};

InstagramAlert.defaultProps = {
  show: false,
  animatedValue: 0.3,
  useNativeDriver: false,
  showCancelButton: false,
  showConfirmButton: false,
  cancelText: 'Cancel',
  confirmText: 'Confirm',
  customView: null,
  modalProps: {},
  cancelButtonTestID: 'instagram-alert-cancel-btn',
  confirmButtonTestID: 'instagram-alert-confirm-btn'
};
