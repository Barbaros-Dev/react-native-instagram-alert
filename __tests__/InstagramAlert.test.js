import 'react-native';
import React from 'react';

import renderer from 'react-test-renderer';

import InstagramAlert from '../src/InstagramAlert';

import props from './props';

test('renders correctly', () => {
  const tree = renderer.create(<InstagramAlert {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});
