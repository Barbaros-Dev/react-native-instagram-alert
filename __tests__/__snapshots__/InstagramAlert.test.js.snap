// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={true}
  visible={true}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flex": 1,
          "height": "100%",
          "justifyContent": "center",
          "position": "absolute",
          "width": "100%",
        },
        undefined,
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "rgba(52,52,52,0.5)",
            "height": 1334,
            "position": "absolute",
            "width": 750,
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 5,
          "maxWidth": "80%",
          "padding": 10,
          "transform": Array [
            Object {
              "scale": 0.30000000000000004,
            },
          ],
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "padding": 10,
            },
            undefined,
          ]
        }
      >
        <ActivityIndicator
          animating={true}
          color="#999999"
          hidesWhenStopped={true}
          size="small"
        />
        <Text
          style={
            Array [
              Object {
                "color": "#626262",
                "fontSize": 18,
                "paddingHorizontal": 15,
                "paddingVertical": 5,
              },
              undefined,
            ]
          }
        >
          Instagram Alert
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#7b7b7b",
                "fontSize": 14,
                "paddingTop": 5,
              },
              undefined,
            ]
          }
        >
          I am just a message
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "flex-end",
              "flexDirection": "row",
              "justifyContent": "center",
              "marginTop": 5,
            },
            undefined,
          ]
        }
      >
        <View
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#D0D0D0",
              "borderRadius": 5,
              "margin": 5,
              "opacity": 1,
              "paddingHorizontal": 10,
              "paddingVertical": 7,
            }
          }
          testID="instagram-alert-cancel-btn"
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontSize": 13,
                },
                undefined,
              ]
            }
          >
            Dismiss
          </Text>
        </View>
        <View
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#AEDEF4",
              "borderRadius": 5,
              "margin": 5,
              "opacity": 1,
              "paddingHorizontal": 10,
              "paddingVertical": 7,
            }
          }
          testID="instagram-alert-confirm-btn"
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontSize": 13,
                },
                undefined,
              ]
            }
          >
            Confirm
          </Text>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
